src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp

```diff
@@ -33,12 +33,20 @@ static std::optional<UInt64> estimateReadRowsCount(QueryPlan::Node & node, bool
             return {};
 
         bool is_filtered_by_index = false;
+        UInt64 total_parts = 0;
+        UInt64 total_granules = 0;
         for (const auto & idx_stat : analyzed_result->index_stats)
         {
+            if (ReadFromMergeTree::IndexType::None == idx_stat.type)
+            {
+                total_parts = idx_stat.num_parts_after;
+                total_granules = idx_stat.num_granules_after;
+                continue;
+            }
             is_filtered_by_index = is_filtered_by_index
-                || (idx_stat.type == ReadFromMergeTree::IndexType::PrimaryKey && !idx_stat.used_keys.empty())
-                || idx_stat.type == ReadFromMergeTree::IndexType::Skip
-                || idx_stat.type == ReadFromMergeTree::IndexType::MinMax;
+                || (total_parts && total_parts < idx_stat.num_parts_after)
```

Comment:
There can be extreme cases, e.g. just 1 out of 1 million granules is eliminated by the index lookup which is equal to not doing an index lookup at all.
Not sure but maybe it makes sense to avoid this by adding a "usefulness" threshold, e.g. 90% of granules are filtered out?

Reply:
The important part is that we get a number of rows close to the estimated one. My assumption was that if we either don’t touch the columns with an index at all (in which case we have no idea how much data is filtered and therefore don’t make any estimation), or if the index is applied, we likely won’t filter much data using this filter at runtime.
Of course, we can have multiple filters or issues with granularity, and it could still be the case. But I’m not sure if a threshold would really help here. What’s more important is that the estimation is lower than the estimation for another table. Perhaps we could set a threshold for that instead.